File: drills/elixir/solutions/nc141_subtree_of_another_tree__by_serialising.exs
defmodule Solution do
  # Serialise both trees and ask whether one string contains the other. That
  # turns an O(n*m) tree comparison into substring search, which is linear with
  # the right algorithm. It is only sound because the serialisation marks the
  # empty children: without them "2" inside "12" would match, and so would a
  # subtree that starts the same way but is missing a child.
  def is_subtree(root, sub), do: String.contains?(serialise(root), serialise(sub))

  defp serialise(nil), do: "#"

  defp serialise({value, left, right}) do
    "(" <> Integer.to_string(value) <> serialise(left) <> serialise(right) <> ")"
  end
end

File: drills/go/solutions/nc141_subtree_of_another_tree__by_serialising.go
package main

import (
	"strconv"
	"strings"
)

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
	// Serialise both with explicit nils and a marker before each value, so
	// a subtree is a substring and "2" cannot match inside "12".
	return strings.Contains(serialise(root), serialise(subRoot))
}

func serialise(node *TreeNode) string {
	if node == nil {
		return ",#"
	}
	return ",^" + strconv.Itoa(node.Val) + serialise(node.Left) + serialise(node.Right)
}
